File: 4.8-bc/ts/stage-3-hexagonal/src/billing/infra/postgres/subscriptionRepo.ts
import type { DbPool } from "../../../shared/db.js";
import type { Repository } from "../../domain/repository.js";
import type { Subscription } from "../../domain/subscription.js";

export class PgSubscriptionRepo implements Repository {
  constructor(private readonly db: DbPool) {}

  async create(s: Subscription): Promise<void> {
    await this.db.query(
      `INSERT INTO billing_subscriptions (id, user_id, plan, next_charge_at, created_at)
       VALUES ($1, $2, $3, $4, $5)`,
      [s.id, s.userId, s.plan, s.nextChargeAt, s.createdAt],
    );
  }
}

File: 4.8-bc/py/stage-3-hexagonal/app/billing/infra/postgres/subscription_repo.py
"""Postgres adapter for Billing Repository."""
from __future__ import annotations

from sqlalchemy import text
from sqlalchemy.ext.asyncio import async_sessionmaker

from app.billing.domain.subscription import Subscription


class SubscriptionRepo:
    def __init__(self, session_maker: async_sessionmaker) -> None:
        self._session_maker = session_maker

    async def create(self, s: Subscription) -> None:
        async with self._session_maker() as session:
            await session.execute(
                text(
                    "INSERT INTO billing_subscriptions "
                    "(id, user_id, plan, next_charge_at, created_at) "
                    "VALUES (:id, :user_id, :plan, :next_charge_at, :created_at)"
                ),
                {
                    "id": s.id,
                    "user_id": s.user_id,
                    "plan": s.plan,
                    "next_charge_at": s.next_charge_at,
                    "created_at": s.created_at,
                },
            )
            await session.commit()
